src/llvm-ptls-reuse.cpp

```diff
@@ -0,0 +1,256 @@
+    }
+    if (refetch_jltls_states_func) {
+        refetch_jltls_states_func->eraseFromParent();
+    }
+    return changed;
+}
+
+void LowerPTLSReuse::propagatePtls(Instruction *Ptls, BbToInstructionMapTy &BbToPtls,
+                                   SmallDenseSet<const Instruction *> &Seen)
+{
+    if (!Seen.insert(Ptls).second) {
+        return;
+    }
+    auto BB = Ptls->getParent();
+    auto InsertResult = BbToPtls.insert({BB, Ptls});
+    if (isa<SelectInst>(Ptls)) {
+        // TODO: Ask if this branch is required. Since we don't introduce select
+        // for PTLS variables explicitly, this won't be necessary if we can be
+        // sure that other LLVM passes won't introduces select instructions.
+        // TODO: Test this branch if we keep this branch.
+        if (!InsertResult.second) {
+            auto OldPtls = InsertResult.first->second;
+            if (OldPtls->comesBefore(Ptls)) {
+                BbToPtls[BB] = Ptls;
+            }
+        }
+    }
+    else if (isa<PHINode>(Ptls)) {
+        // Nothing to do, since PHINode is always before other instructions.
+    }
+    else {
+        return;
+    }
+    for (auto User : Ptls->users()) {
+        if (auto I = dyn_cast<Instruction>(User)) {
+            propagatePtls(I, BbToPtls, Seen);
+        }
+    }
+}
+
+Instruction *LowerPTLSReuse::mergePtls(BasicBlock *BB, BbToInstructionMapTy &BbToPtls)
+{
+    if (BB == &BB->getParent()->getEntryBlock()) {
+        return ensureEntryBlockPtls(*BB->getParent());
+    }
+    SmallVector<BasicBlock *, 2> PreBbs;
+    for (auto Pre : predecessors(BB)) {
+        PreBbs.push_back(Pre);
+    }
+    auto Phi = PHINode::Create(PointerType::get(T_ppjlvalue, 0), PreBbs.size(), "ptls",
+                               BB->getFirstNonPHI() ? BB->getFirstNonPHI() : &BB->front());
+    if (!BbToPtls[BB]) {
+        // PTLS from this BB is requested as a transitive dependency. So,
+        // marking the phi node as the PTLS at the end of this basic block. This
+        // has to be done before the recursions.
+        BbToPtls[BB] = Phi;
+        assert(Phi->getType() == PtlsType);
+    }
+    for (auto Pre : PreBbs) {
+        Phi->addIncoming(ptlsForBb(Pre, BbToPtls), Pre);
+    }
+    return Phi;
+}
+
+// Get PTLS usable at the end of `BB`.
+Instruction *LowerPTLSReuse::ptlsForBb(BasicBlock *BB, BbToInstructionMapTy &BbToPtls)
+{
+    if (auto Ptls = BbToPtls[BB]) {
+        return Ptls;
+    }
+    auto Ptls = mergePtls(BB, BbToPtls);
+    assert(Ptls->getType() == PtlsType);
+    return Ptls;
+}
+
+void LowerPTLSReuse::populatePtls(BasicBlock *BB, BbToInstructionMapTy &BbToPtls)
+{
+    // First, go over instructions in `BB` and construct a mapping between
+    // "source" PTLS and its reuses.
+    Instruction *Ptls = mergePtls(BB, BbToPtls);
+    SmallVector<CallInst *> Refetches;
+    SmallVector<std::pair<Instruction *, std::unique_ptr<SmallVector<Instruction *>>>>
+        SourceAndReuses;
```

Comment:
I have no idea how to do `SmallVector`-of-`SmallVector`s in a "modern" C++ (or maybe rather in LLVM infrastructure). Is this an OK approach?